src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -16,64 +16,319 @@
+        }
+
+        return measurements;
     }
+
+    /**
+     * -------- This section contains SWAR code (SIMD Within A Register) which processes a bytebuffer as longs to find values:
+     */
+
+    private int findNextSWAR(ByteBuffer bb, long pattern, int start, int limit) {
+        int i;
+        for (i = start; i <= limit - 8; i += 8) {
+            long word = bb.getLong(i);
+            int index = firstAnyPattern(word, pattern);
+            if (index < Long.BYTES) {
+                return i + index;
+            }
+        }
+        // Handle remaining bytes
+        for (; i < limit; i++) {
+            if (bb.get(i) == (byte) pattern) {
+                return i;
+            }
+        }
+        return limit; // delimiter not found
+    }
+
+    private static long compilePattern(byte value) {
+        return ((long) value << 56) | ((long) value << 48) | ((long) value << 40) | ((long) value << 32) |
+                ((long) value << 24) | ((long) value << 16) | ((long) value << 8) | (long) value;
+    }
+
+    private static int firstAnyPattern(long word, long pattern) {
+        final long match = word ^ pattern;
+        long mask = match - 0x0101010101010101L;
+        mask &= ~match;
+        mask &= 0x8080808080808080L;
+        return Long.numberOfLeadingZeros(mask) >> 3;
+    }
+
+    /**
+     * -------- Thanks to bjhara for the idea of using memory mapped files.
+     * @param fileChannel
+     * @return
+     * @throws IOException
+     */
+    private static Stream<ByteBuffer> splitFileChannel(final FileChannel fileChannel) throws IOException {
+        return StreamSupport.stream(Spliterators.spliteratorUnknownSize(new Iterator<>() {
+            private static final long CHUNK_SIZE = (long) Math.pow(2, 19);
+
+            private final long size = fileChannel.size();
+            private long bytesRead = 0;
+
+            @Override
+            public ByteBuffer next() {
+                try {
+                    final MappedByteBuffer mappedByteBuffer = fileChannel.map(FileChannel.MapMode.READ_ONLY, bytesRead, Math.min(CHUNK_SIZE, size - bytesRead));
+
+                    // Adjust end to start of a line:
+                    int realEnd = mappedByteBuffer.limit() - 1;
+                    while (mappedByteBuffer.get(realEnd) != '\n') {
+                        realEnd--;
+                    }
+                    mappedByteBuffer.limit(++realEnd);
+                    bytesRead += realEnd;
+
+                    return mappedByteBuffer;
+                }
+                catch (IOException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            @Override
+            public boolean hasNext() {
+                return bytesRead < size;
+            }
+        }, Spliterator.IMMUTABLE), false);
+    }
+
+    /**
+     * Branchless parser, goes from String to int (10x):
+     * "-1.2" to -12
+     * "40.1" to 401
```

Comment:
@gunnarmorling Is this assumption acceptable? E.g. `1`, `0`, `2.00` are all valid doubles.

Reply:
Also there should be an acceptance test suite for the implementations, I am pretty sure this implementation does not produce the same output as the baseline.

Reply:
I'm pretty sure it produces the exact same output, I check regularly with each change.

The input and output have one decimal place precision (as stated by the website "rounded to one fractional digit").

I'm internally storing the doubles as 10x integers because the precision is just a single digit, and I'm making sure the rounding it correct afterwards for the average.

Reply:
The README only says about output format but not input.
https://github.com/gunnarmorling/1brc/blob/e7e7deb64a9d1ace22e5f9091ba7c06d1d7f1cd4/README.md?plain=1#L27-L28
so its worth clarifying.

There are many rounding modes - this is also not specified, e.g. in go https://pkg.go.dev/math#Round is not the same as in java.

Reply:
Also reference implementation does ad-hoc rounding `Math.round(value * 10.0) / 10.0` but actual output depends on the string concatenation which performs another rounding, see https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Double.html#toString(double)

Reply:
> I am pretty sure this implementation does not produce the same output as the baseline.
> I'm pretty sure it produces the exact same output, I check regularly with each change.

To prove the point here an input https://gist.github.com/AlexanderYastrebov/d560d697c61484cc5eedaa09120d6464 for which your implementation produces different output than the baseline.

Due to unfortunate output format (see https://github.com/gunnarmorling/1brc/issues/14) one hs to use word diff:
```diff
$ git --no-pager diff --word-diff /tmp/baseline.txt /tmp/royvanrijn.txt
diff --git a/tmp/baseline.txt b/tmp/royvanrijn.txt
index 1a3d2ea..5c739b7 100644
--- a/tmp/baseline.txt
+++ b/tmp/royvanrijn.txt
@@ -1 +1 @@
<omitted for brevity>, Yakutsk=3.4/3.4/3.4, [-Zagreb=7.7/12.5/17.4}-]{+Zagreb=7.7/12.6/17.4}+}
```

You can find such inputs yourself via
```sh
$ while ./create_measurements.sh 100 && diff <(./calculate_average_royvanrijn.sh) <(./calculate_av
erage.sh) ; do echo OK; done
```

Reply:
> Also reference implementation does ad-hoc rounding Math.round(value * 10.0) / 10.0 but actual output depends on the string concatenation which performs another rounding

The baseline rounding is not always correct as could be shown in java playground https://dev.java/playground/ via
```java
double value = (33.6 + 31.7 + 21.9 + 14.6) / 4.0; // 101.8/4 == 25.45 exactly but can not be represented as double; it is 25.449999999999996
double rounded = Math.round(value * 10.0) / 10.0; // 25.45 * 10 == 254.5, round(254.5) == 255, 255/10 should be 25.5

System.out.println("==" + rounded + "=="); // but prints ==25.4==
```

Reply:
Ah, you are correct in that example, my code would correctly return 25.5 in that last one instead of 25.4.
That's a bit of a pain...

Let's see what @gunnarmorling says about the rounding.

Reply:
> The README only says about output format but not input

Ah yes. For the sake of simplicity, it's to assume that each input value also has exactly one fractional digit. I'll clarify that in the README.

> Let's see what @gunnarmorling says about the rounding.

Sorry for not being explicit enough here. Can the behavior of the reference implementation be described using any of the existing values of [RoundingMode](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/math/RoundingMode.html)? E.g. `FLOOR`? If so, I'd suggest we make this explicit and any implementation would have to conform to that.

Reply:
But that's the thing, I don't think we can change the behavior of the reference implementation at this point, as it would render existing submissions invalid if they implement a different behavior. So I'd rather make the behavior of the RI explicit, also if it's not the most natural one (agreed that `HALF_UP` behavior would have been better).

Reply:
I don't think its possible to fix RI because it uses double division and rounds twice.
Since there are no acceptance tests I bet a lot of implementations (those that do not parse and calculate values the same way) will not match RI anyways.

I think RI should favor correctness over performance, then it can be used to build acceptance test suite.

Reply:
Ok, I've logged https://github.com/gunnarmorling/1brc/issues/49 for getting this one sorted out separately and get this PR merged. Let's continue the rounding topic over there. Thx!

Reply:
I find that `Math.round((double) x / y) / 10.0` produces the same result as the reference - the important part is to do the division in `double` - addition is guaranteed to be loseless.